src/allreduce_robust.h

```diff
@@ -632,6 +633,14 @@ o   *  the input state must exactly one saved state(local state of current node)
   int local_chkpt_version;
   // if checkpoint were loaded, used to distinguish results boostrap cache from seqno cache
   bool checkpoint_loaded;
+  // sidecar executing timeout task
+  std::future<bool> rabit_timeout_task;
+  // flag to shutdown rabit_timeout_task before timeout
+  std::atomic<bool> shutdown_timeout{false};
+  // error handler
+  void (* _error)(const char *fmt, ...) = utils::Error;
+  // assert handler
+  void (* _assert)(bool exp, const char *fmt, ...) = utils::Assert;
```

Comment:
I see.  But why?

Reply:
I will jump into this when time allows.

Reply:
something like this https://github.com/chenqin/rabit/commit/dba2836eb9c63655de0351d19215b56dc162d0b3 (WIP)